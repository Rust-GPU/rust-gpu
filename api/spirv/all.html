<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="List of all items in this crate"><title>List of all items in this crate</title><script>if(window.location.protocol!=="file:")document.head.insertAdjacentHTML("beforeend","SourceSerif4-Regular-46f98efaafac5295.ttf.woff2,FiraSans-Regular-018c141bf0843ffd.woff2,FiraSans-Medium-8f9a781e4970d388.woff2,SourceCodePro-Regular-562dcc5011b6de7d.ttf.woff2,SourceCodePro-Semibold-d899c5a5c4aeb14a.ttf.woff2".split(",").map(f=>`<link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/${f}">`).join(""))</script><link rel="stylesheet" href="../static.files/normalize-76eba96aa4d2e634.css"><link rel="stylesheet" href="../static.files/rustdoc-081576b923113409.css"><meta name="rustdoc-vars" data-root-path="../" data-static-root-path="../static.files/" data-current-crate="spirv" data-themes="" data-resource-suffix="" data-rustdoc-version="1.79.0-nightly (244da22fa 2024-04-23)" data-channel="nightly" data-search-js="search-bf21c90c8c1d92b1.js" data-settings-js="settings-4313503d2e1961c2.js" ><script src="../static.files/storage-e32f0c247825364d.js"></script><script defer src="../static.files/main-20a3ad099b048cf2.js"></script><noscript><link rel="stylesheet" href="../static.files/noscript-09095024cf37855e.css"></noscript><link rel="alternate icon" type="image/png" href="../static.files/favicon-32x32-422f7d1d52889060.png"><link rel="icon" type="image/svg+xml" href="../static.files/favicon-2c020d218678b618.svg"></head><body class="rustdoc mod sys"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="mobile-topbar"><button class="sidebar-menu-toggle" title="show sidebar"></button></nav><nav class="sidebar"><div class="sidebar-crate"><h2><a href="../spirv/index.html">spirv</a><span class="version">0.3.0+sdk-1.3.268.0</span></h2></div><div class="sidebar-elems"><section><ul class="block"><li><a href="#structs">Structs</a></li><li><a href="#enums">Enums</a></li><li><a href="#constants">Constants</a></li><li><a href="#types">Type Aliases</a></li></ul></section></div></nav><div class="sidebar-resizer"></div><main><div class="width-limiter"><nav class="sub"><form class="search-form"><span></span><div id="sidebar-button" tabindex="-1"><a href="../spirv/all.html" title="show sidebar"></a></div><input class="search-input" name="search" aria-label="Run search in the documentation" autocomplete="off" spellcheck="false" placeholder="Type ‘S’ or ‘/’ to search, ‘?’ for more options…" type="search"><div id="help-button" tabindex="-1"><a href="../help.html" title="help">?</a></div><div id="settings-menu" tabindex="-1"><a href="../settings.html" title="settings">Settings</a></div></form></nav><section id="main-content" class="content"><h1>List of all items</h1><h3 id="structs">Structs</h3><ul class="all-items"><li><a href="struct.CooperativeMatrixOperands.html">CooperativeMatrixOperands</a></li><li><a href="struct.FPFastMathMode.html">FPFastMathMode</a></li><li><a href="struct.FragmentShadingRate.html">FragmentShadingRate</a></li><li><a href="struct.FunctionControl.html">FunctionControl</a></li><li><a href="struct.ImageOperands.html">ImageOperands</a></li><li><a href="struct.KernelProfilingInfo.html">KernelProfilingInfo</a></li><li><a href="struct.LoopControl.html">LoopControl</a></li><li><a href="struct.MemoryAccess.html">MemoryAccess</a></li><li><a href="struct.MemorySemantics.html">MemorySemantics</a></li><li><a href="struct.RayFlags.html">RayFlags</a></li><li><a href="struct.SelectionControl.html">SelectionControl</a></li></ul><h3 id="enums">Enums</h3><ul class="all-items"><li><a href="enum.AccessQualifier.html">AccessQualifier</a></li><li><a href="enum.AddressingModel.html">AddressingModel</a></li><li><a href="enum.BuiltIn.html">BuiltIn</a></li><li><a href="enum.CLOp.html">CLOp</a></li><li><a href="enum.Capability.html">Capability</a></li><li><a href="enum.CooperativeMatrixLayout.html">CooperativeMatrixLayout</a></li><li><a href="enum.CooperativeMatrixUse.html">CooperativeMatrixUse</a></li><li><a href="enum.Decoration.html">Decoration</a></li><li><a href="enum.Dim.html">Dim</a></li><li><a href="enum.ExecutionMode.html">ExecutionMode</a></li><li><a href="enum.ExecutionModel.html">ExecutionModel</a></li><li><a href="enum.FPDenormMode.html">FPDenormMode</a></li><li><a href="enum.FPOperationMode.html">FPOperationMode</a></li><li><a href="enum.FPRoundingMode.html">FPRoundingMode</a></li><li><a href="enum.FunctionParameterAttribute.html">FunctionParameterAttribute</a></li><li><a href="enum.GLOp.html">GLOp</a></li><li><a href="enum.GroupOperation.html">GroupOperation</a></li><li><a href="enum.HostAccessQualifier.html">HostAccessQualifier</a></li><li><a href="enum.ImageChannelDataType.html">ImageChannelDataType</a></li><li><a href="enum.ImageChannelOrder.html">ImageChannelOrder</a></li><li><a href="enum.ImageFormat.html">ImageFormat</a></li><li><a href="enum.InitializationModeQualifier.html">InitializationModeQualifier</a></li><li><a href="enum.KernelEnqueueFlags.html">KernelEnqueueFlags</a></li><li><a href="enum.LinkageType.html">LinkageType</a></li><li><a href="enum.LoadCacheControl.html">LoadCacheControl</a></li><li><a href="enum.MemoryModel.html">MemoryModel</a></li><li><a href="enum.Op.html">Op</a></li><li><a href="enum.OverflowModes.html">OverflowModes</a></li><li><a href="enum.PackedVectorFormat.html">PackedVectorFormat</a></li><li><a href="enum.QuantizationModes.html">QuantizationModes</a></li><li><a href="enum.RayQueryCandidateIntersectionType.html">RayQueryCandidateIntersectionType</a></li><li><a href="enum.RayQueryCommittedIntersectionType.html">RayQueryCommittedIntersectionType</a></li><li><a href="enum.RayQueryIntersection.html">RayQueryIntersection</a></li><li><a href="enum.SamplerAddressingMode.html">SamplerAddressingMode</a></li><li><a href="enum.SamplerFilterMode.html">SamplerFilterMode</a></li><li><a href="enum.Scope.html">Scope</a></li><li><a href="enum.SourceLanguage.html">SourceLanguage</a></li><li><a href="enum.StorageClass.html">StorageClass</a></li><li><a href="enum.StoreCacheControl.html">StoreCacheControl</a></li></ul><h3 id="types">Type Aliases</h3><ul class="all-items"><li><a href="type.Word.html">Word</a></li></ul><h3 id="constants">Constants</h3><ul class="all-items"><li><a href="constant.MAGIC_NUMBER.html">MAGIC_NUMBER</a></li><li><a href="constant.MAJOR_VERSION.html">MAJOR_VERSION</a></li><li><a href="constant.MINOR_VERSION.html">MINOR_VERSION</a></li><li><a href="constant.REVISION.html">REVISION</a></li></ul></section></div></main></body></html>