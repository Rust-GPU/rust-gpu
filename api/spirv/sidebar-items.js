window.SIDEBAR_ITEMS = {"constant":["MAGIC_NUMBER","MAJOR_VERSION","MINOR_VERSION","REVISION"],"enum":["AccessQualifier","AddressingModel","BuiltIn","CLOp","Capability","CooperativeMatrixLayout","CooperativeMatrixUse","Decoration","Dim","ExecutionMode","ExecutionModel","FPDenormMode","FPOperationMode","FPRoundingMode","FunctionParameterAttribute","GLOp","GroupOperation","HostAccessQualifier","ImageChannelDataType","ImageChannelOrder","ImageFormat","InitializationModeQualifier","KernelEnqueueFlags","LinkageType","LoadCacheControl","MemoryModel","Op","OverflowModes","PackedVectorFormat","QuantizationModes","RayQueryCandidateIntersectionType","RayQueryCommittedIntersectionType","RayQueryIntersection","SamplerAddressingMode","SamplerFilterMode","Scope","SourceLanguage","StorageClass","StoreCacheControl"],"struct":["CooperativeMatrixOperands","FPFastMathMode","FragmentShadingRate","FunctionControl","ImageOperands","KernelProfilingInfo","LoopControl","MemoryAccess","MemorySemantics","RayFlags","SelectionControl"],"type":["Word"]};