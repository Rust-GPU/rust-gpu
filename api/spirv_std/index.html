<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="`spirv-std`"><title>spirv_std - Rust</title><script>if(window.location.protocol!=="file:")document.head.insertAdjacentHTML("beforeend","SourceSerif4-Regular-46f98efaafac5295.ttf.woff2,FiraSans-Regular-018c141bf0843ffd.woff2,FiraSans-Medium-8f9a781e4970d388.woff2,SourceCodePro-Regular-562dcc5011b6de7d.ttf.woff2,SourceCodePro-Semibold-d899c5a5c4aeb14a.ttf.woff2".split(",").map(f=>`<link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/${f}">`).join(""))</script><link rel="stylesheet" href="../static.files/normalize-76eba96aa4d2e634.css"><link rel="stylesheet" href="../static.files/rustdoc-081576b923113409.css"><meta name="rustdoc-vars" data-root-path="../" data-static-root-path="../static.files/" data-current-crate="spirv_std" data-themes="" data-resource-suffix="" data-rustdoc-version="1.79.0-nightly (244da22fa 2024-04-23)" data-channel="nightly" data-search-js="search-bf21c90c8c1d92b1.js" data-settings-js="settings-4313503d2e1961c2.js" ><script src="../static.files/storage-e32f0c247825364d.js"></script><script defer src="../crates.js"></script><script defer src="../static.files/main-20a3ad099b048cf2.js"></script><noscript><link rel="stylesheet" href="../static.files/noscript-09095024cf37855e.css"></noscript><link rel="alternate icon" type="image/png" href="../static.files/favicon-32x32-422f7d1d52889060.png"><link rel="icon" type="image/svg+xml" href="../static.files/favicon-2c020d218678b618.svg"></head><body class="rustdoc mod crate"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="mobile-topbar"><button class="sidebar-menu-toggle" title="show sidebar"></button></nav><nav class="sidebar"><div class="sidebar-crate"><h2><a href="../spirv_std/index.html">spirv_std</a><span class="version">0.9.0</span></h2></div><div class="sidebar-elems"><ul class="block"><li><a id="all-types" href="all.html">All Items</a></li></ul><section><ul class="block"><li><a href="#reexports">Re-exports</a></li><li><a href="#modules">Modules</a></li><li><a href="#macros">Macros</a></li><li><a href="#structs">Structs</a></li><li><a href="#attributes">Attribute Macros</a></li></ul></section></div></nav><div class="sidebar-resizer"></div><main><div class="width-limiter"><nav class="sub"><form class="search-form"><span></span><div id="sidebar-button" tabindex="-1"><a href="../spirv_std/all.html" title="show sidebar"></a></div><input class="search-input" name="search" aria-label="Run search in the documentation" autocomplete="off" spellcheck="false" placeholder="Type ‘S’ or ‘/’ to search, ‘?’ for more options…" type="search"><div id="help-button" tabindex="-1"><a href="../help.html" title="help">?</a></div><div id="settings-menu" tabindex="-1"><a href="../settings.html" title="settings">Settings</a></div></form></nav><section id="main-content" class="content"><div class="main-heading"><h1>Crate <a class="mod" href="#">spirv_std</a><button id="copy-path" title="Copy item path to clipboard">Copy item path</button></h1><span class="out-of-band"><a class="src" href="../src/spirv_std/lib.rs.html#1-152">source</a> · <button id="toggle-all-docs" title="collapse all docs">[<span>&#x2212;</span>]</button></span></div><details class="toggle top-doc" open><summary class="hideme"><span>Expand description</span></summary><div class="docblock"><h2 id="spirv-std"><a class="doc-anchor" href="#spirv-std">§</a><code>spirv-std</code></h2>
<p>Core functions, traits, and more that make up a “standard library” for SPIR-V for use in <a href="https://github.com/rust-gpu/rust-gpu#readme">rust-gpu</a>.</p>
<p>This crate gives a <code>rust-gpu</code> shader access to the required <code>#![spirv(..)]</code> attribute, as well as povide all kinds of APIs that allows a shader to access GPU resources such as textures and buffers. Optionally, through the use of the <code>&quot;glam&quot;</code> feature, it includes some boilerplate trait implementations to make <code>glam</code> vector types compatible with these APIs.</p>
<h3 id="example"><a class="doc-anchor" href="#example">§</a>Example</h3>
<p><img src="https://github.com/rust-gpu/rust-gpu/raw/b12a2f3f6a54bc841d05a9224bc577909d519228/docs/assets/sky.jpg" alt="Sky shader" /></p>
<p>Here is a small excerpt to see what a shader would look like. See <a href="https://github.com/rust-gpu/rust-gpu/blob/69cb69d28f1e64420ee31ade5e7dffb7c5621e89/examples/shaders/sky-shader/src/lib.rs">source</a> for full details of the shader that generates above image.</p>

<div class="example-wrap ignore"><a href="#" class="tooltip" title="This example is not tested">ⓘ</a><pre class="rust rust-example-rendered"><code><span class="kw">use </span>spirv_std::spirv;
<span class="kw">use </span>glam::{Vec3, Vec4, vec2, vec3};

<span class="attr">#[spirv(fragment)]
</span><span class="kw">pub fn </span>main(
    <span class="attr">#[spirv(frag_coord)] </span>in_frag_coord: <span class="kw-2">&amp;</span>Vec4,
    <span class="attr">#[spirv(push_constant)] </span>constants: <span class="kw-2">&amp;</span>ShaderConstants,
    output: <span class="kw-2">&amp;mut </span>Vec4,
) {
    <span class="kw">let </span>frag_coord = vec2(in_frag_coord.x, in_frag_coord.y);
    <span class="kw">let </span><span class="kw-2">mut </span>uv = (frag_coord - <span class="number">0.5 </span>* vec2(constants.width <span class="kw">as </span>f32, constants.height <span class="kw">as </span>f32))
        / constants.height <span class="kw">as </span>f32;
    uv.y = -uv.y;

    <span class="kw">let </span>eye_pos = vec3(<span class="number">0.0</span>, <span class="number">0.0997</span>, <span class="number">0.2</span>);
    <span class="kw">let </span>sun_pos = vec3(<span class="number">0.0</span>, <span class="number">75.0</span>, -<span class="number">1000.0</span>);
    <span class="kw">let </span>dir = get_ray_dir(uv, eye_pos, sun_pos);

    <span class="comment">// evaluate Preetham sky model
    </span><span class="kw">let </span>color = sky(dir, sun_pos);

    <span class="kw-2">*</span>output = tonemap(color).extend(<span class="number">1.0</span>)
}</code></pre></div>
<h3 id="getting-started"><a class="doc-anchor" href="#getting-started">§</a>Getting started</h3>
<p>Check out <a href="https://rust-gpu.github.io/rust-gpu/book/">The <code>rust-gpu</code> Dev Guide</a> for information on how to get started with using it in your projects.</p>
<p>Experiment with rust-gpu shaders in-browser at <a href="https://shadered.org/shaders?language=rust&amp;sort=hot">SHADERed</a>.</p>
</div></details><h2 id="reexports" class="section-header">Re-exports<a href="#reexports" class="anchor">§</a></h2><ul class="item-table"><li><div class="item-name"><code>pub extern crate <a class="mod" href="../spirv_std_macros/index.html" title="mod spirv_std_macros">spirv_std_macros</a> as macros;</code></div></li><li><div class="item-name" id="reexport.ByteAddressableBuffer"><code>pub use byte_addressable_buffer::<a class="struct" href="byte_addressable_buffer/struct.ByteAddressableBuffer.html" title="struct spirv_std::byte_addressable_buffer::ByteAddressableBuffer">ByteAddressableBuffer</a>;</code></div></li><li><div class="item-name" id="reexport.num_traits"><code>pub use <a class="mod" href="../num_traits/index.html" title="mod num_traits">num_traits</a>;</code></div></li><li><div class="item-name" id="reexport.glam"><code>pub use <a class="mod" href="../glam/index.html" title="mod glam">glam</a>;</code></div></li></ul><h2 id="modules" class="section-header">Modules<a href="#modules" class="anchor">§</a></h2><ul class="item-table"><li><div class="item-name"><a class="mod" href="arch/index.html" title="mod spirv_std::arch">arch</a></div><div class="desc docblock-short">SPIR-V Intrinsics</div></li><li><div class="item-name"><a class="mod" href="byte_addressable_buffer/index.html" title="mod spirv_std::byte_addressable_buffer">byte_addressable_buffer</a></div><div class="desc docblock-short">Container for an untyped blob of data.</div></li><li><div class="item-name"><a class="mod" href="float/index.html" title="mod spirv_std::float">float</a></div><div class="desc docblock-short">Traits and helper functions related to floats.</div></li><li><div class="item-name"><a class="mod" href="image/index.html" title="mod spirv_std::image">image</a></div><div class="desc docblock-short">Image types</div></li><li><div class="item-name"><a class="mod" href="indirect_command/index.html" title="mod spirv_std::indirect_command">indirect_command</a></div><div class="desc docblock-short">Indirect command structs from vulkan</div></li><li><div class="item-name"><a class="mod" href="integer/index.html" title="mod spirv_std::integer">integer</a></div><div class="desc docblock-short">Traits related to integers.</div></li><li><div class="item-name"><a class="mod" href="memory/index.html" title="mod spirv_std::memory">memory</a></div><div class="desc docblock-short">Types for handling memory ordering constraints for concurrent memory access.</div></li><li><div class="item-name"><a class="mod" href="number/index.html" title="mod spirv_std::number">number</a></div><div class="desc docblock-short">Traits and helper functions related to numbers.</div></li><li><div class="item-name"><a class="mod" href="ray_tracing/index.html" title="mod spirv_std::ray_tracing">ray_tracing</a></div><div class="desc docblock-short">Ray-tracing data types</div></li><li><div class="item-name"><a class="mod" href="scalar/index.html" title="mod spirv_std::scalar">scalar</a></div><div class="desc docblock-short">Traits related to scalars.</div></li><li><div class="item-name"><a class="mod" href="vector/index.html" title="mod spirv_std::vector">vector</a></div><div class="desc docblock-short">Traits related to vectors.</div></li></ul><h2 id="macros" class="section-header">Macros<a href="#macros" class="anchor">§</a></h2><ul class="item-table"><li><div class="item-name"><a class="macro" href="macro.Image.html" title="macro spirv_std::Image">Image</a></div><div class="desc docblock-short">A macro for creating SPIR-V <code>OpTypeImage</code> types. Always produces a
<code>spirv_std::image::Image&lt;...&gt;</code> type.</div></li><li><div class="item-name"><a class="macro" href="macro.ray_query.html" title="macro spirv_std::ray_query">ray_query</a></div><div class="desc docblock-short">Constructs an uninitialized ray query variable. Using the syntax
<code>let (mut)? &lt;name&gt;</code>. Where <code>name</code> is the name of the ray query variable.</div></li></ul><h2 id="structs" class="section-header">Structs<a href="#structs" class="anchor">§</a></h2><ul class="item-table"><li><div class="item-name"><a class="struct" href="struct.RuntimeArray.html" title="struct spirv_std::RuntimeArray">RuntimeArray</a></div><div class="desc docblock-short">Dynamically-sized arrays in Rust carry around their length as the second half of a tuple.
Unfortunately, sometimes SPIR-V provides an unsized array with no way of obtaining its length.
Hence, this type represents something very similar to a slice, but with no way of knowing its
length.</div></li><li><div class="item-name"><a class="struct" href="struct.Sampler.html" title="struct spirv_std::Sampler">Sampler</a></div><div class="desc docblock-short">An opaque reference to settings that describe how to access, filter, or
sample an image.</div></li><li><div class="item-name"><a class="struct" href="struct.TypedBuffer.html" title="struct spirv_std::TypedBuffer">TypedBuffer</a></div><div class="desc docblock-short">Explicit (uniform/storage) buffer handle for descriptor indexing.</div></li></ul><h2 id="attributes" class="section-header">Attribute Macros<a href="#attributes" class="anchor">§</a></h2><ul class="item-table"><li><div class="item-name"><a class="attr" href="attr.spirv.html" title="attr spirv_std::spirv">spirv</a></div><div class="desc docblock-short">Replaces all (nested) occurrences of the <code>#[spirv(..)]</code> attribute with
<code>#[cfg_attr(target_arch=&quot;spirv&quot;, rust_gpu::spirv(..))]</code>.</div></li></ul></section></div></main></body></html>