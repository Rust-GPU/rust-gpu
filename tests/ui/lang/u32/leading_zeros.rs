// Test all trailing and leading zeros. No need to test ones, they just call the zero variant with !value

// build-pass

use spirv_std::spirv;

#[spirv(fragment)]
pub fn leading_zeros_u8(
    #[spirv(descriptor_set = 0, binding = 0, storage_buffer)] buffer: &u8,
    out: &mut u32,
) {
    *out = u8::leading_zeros(*buffer);
}

#[spirv(fragment)]
pub fn leading_zeros_u16(
    #[spirv(descriptor_set = 0, binding = 0, storage_buffer)] buffer: &u16,
    out: &mut u32,
) {
    *out = u16::leading_zeros(*buffer);
}

#[spirv(fragment)]
pub fn leading_zeros_u32(
    #[spirv(descriptor_set = 0, binding = 0, storage_buffer)] buffer: &u32,
    out: &mut u32,
) {
    *out = u32::leading_zeros(*buffer);
}

#[spirv(fragment)]
pub fn leading_zeros_u64(
    #[spirv(descriptor_set = 0, binding = 0, storage_buffer)] buffer: &u64,
    out: &mut u32,
) {
    *out = u64::leading_zeros(*buffer);
}

#[spirv(fragment)]
pub fn leading_zeros_i8(
    #[spirv(descriptor_set = 0, binding = 0, storage_buffer)] buffer: &i8,
    out: &mut u32,
) {
    *out = i8::leading_zeros(*buffer);
}

#[spirv(fragment)]
pub fn leading_zeros_i16(
    #[spirv(descriptor_set = 0, binding = 0, storage_buffer)] buffer: &i16,
    out: &mut u32,
) {
    *out = i16::leading_zeros(*buffer);
}

#[spirv(fragment)]
pub fn leading_zeros_i32(
    #[spirv(descriptor_set = 0, binding = 0, storage_buffer)] buffer: &i32,
    out: &mut u32,
) {
    *out = i32::leading_zeros(*buffer);
}

#[spirv(fragment)]
pub fn leading_zeros_i64(
    #[spirv(descriptor_set = 0, binding = 0, storage_buffer)] buffer: &i64,
    out: &mut u32,
) {
    *out = i64::leading_zeros(*buffer);
}
