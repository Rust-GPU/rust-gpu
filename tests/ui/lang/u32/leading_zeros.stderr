error: cannot cast between pointer types
       from `*u32`
         to `*u8`
   --> /home/firestar99/.rustup/toolchains/nightly-2024-11-22-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/num/mod.rs:535:5
    |
535 | /     uint_impl! {
536 | |         Self = u8,
537 | |         ActualT = u8,
538 | |         SignedT = i8,
...   |
552 | |         bound_condition = "",
553 | |     }
    | |_____^
    |
note: used from within `<u8>::leading_zeros`
   --> /home/firestar99/.rustup/toolchains/nightly-2024-11-22-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/num/mod.rs:535:5
    |
535 | /     uint_impl! {
536 | |         Self = u8,
537 | |         ActualT = u8,
538 | |         SignedT = i8,
...   |
552 | |         bound_condition = "",
553 | |     }
    | |_____^
note: called by `leading_zeros::leading_zeros_u8`
   --> $DIR/leading_zeros.rs:12:12
    |
12  |     *out = u8::leading_zeros(*buffer);
    |            ^^^^^^^^^^^^^^^^^^^^^^^^^^
note: called by `leading_zeros_u8`
   --> $DIR/leading_zeros.rs:8:8
    |
8   | pub fn leading_zeros_u8(
    |        ^^^^^^^^^^^^^^^^

error: cannot cast between pointer types
       from `*u32`
         to `*u16`
    --> /home/firestar99/.rustup/toolchains/nightly-2024-11-22-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/num/mod.rs:1151:5
     |
1151 | /     uint_impl! {
1152 | |         Self = u16,
1153 | |         ActualT = u16,
1154 | |         SignedT = i16,
...    |
1168 | |         bound_condition = "",
1169 | |     }
     | |_____^
     |
note: used from within `<u16>::leading_zeros`
    --> /home/firestar99/.rustup/toolchains/nightly-2024-11-22-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/num/mod.rs:1151:5
     |
1151 | /     uint_impl! {
1152 | |         Self = u16,
1153 | |         ActualT = u16,
1154 | |         SignedT = i16,
...    |
1168 | |         bound_condition = "",
1169 | |     }
     | |_____^
note: called by `leading_zeros::leading_zeros_u16`
    --> $DIR/leading_zeros.rs:20:12
     |
20   |     *out = u16::leading_zeros(*buffer);
     |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^
note: called by `leading_zeros_u16`
    --> $DIR/leading_zeros.rs:16:8
     |
16   | pub fn leading_zeros_u16(
     |        ^^^^^^^^^^^^^^^^^

error: cannot cast between pointer types
       from `*u32`
         to `*u64`
    --> /home/firestar99/.rustup/toolchains/nightly-2024-11-22-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/num/mod.rs:1223:5
     |
1223 | /     uint_impl! {
1224 | |         Self = u64,
1225 | |         ActualT = u64,
1226 | |         SignedT = i64,
...    |
1240 | |         bound_condition = "",
1241 | |     }
     | |_____^
     |
note: used from within `<u64>::leading_zeros`
    --> /home/firestar99/.rustup/toolchains/nightly-2024-11-22-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/num/mod.rs:1223:5
     |
1223 | /     uint_impl! {
1224 | |         Self = u64,
1225 | |         ActualT = u64,
1226 | |         SignedT = i64,
...    |
1240 | |         bound_condition = "",
1241 | |     }
     | |_____^
note: called by `leading_zeros::leading_zeros_u64`
    --> $DIR/leading_zeros.rs:36:12
     |
36   |     *out = u64::leading_zeros(*buffer);
     |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^
note: called by `leading_zeros_u64`
    --> $DIR/leading_zeros.rs:32:8
     |
32   | pub fn leading_zeros_u64(
     |        ^^^^^^^^^^^^^^^^^

error: cannot cast between pointer types
       from `*u32`
         to `*u8`
   --> /home/firestar99/.rustup/toolchains/nightly-2024-11-22-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/num/mod.rs:535:5
    |
535 | /     uint_impl! {
536 | |         Self = u8,
537 | |         ActualT = u8,
538 | |         SignedT = i8,
...   |
552 | |         bound_condition = "",
553 | |     }
    | |_____^
    |
note: used from within `<u8>::leading_zeros`
   --> /home/firestar99/.rustup/toolchains/nightly-2024-11-22-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/num/mod.rs:535:5
    |
535 | /     uint_impl! {
536 | |         Self = u8,
537 | |         ActualT = u8,
538 | |         SignedT = i8,
...   |
552 | |         bound_condition = "",
553 | |     }
    | |_____^
note: called by `<i8>::leading_zeros`
   --> /home/firestar99/.rustup/toolchains/nightly-2024-11-22-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/num/mod.rs:335:5
    |
335 | /     int_impl! {
336 | |         Self = i8,
337 | |         ActualT = i8,
338 | |         UnsignedT = u8,
...   |
353 | |         bound_condition = "",
354 | |     }
    | |_____^
note: called by `leading_zeros::leading_zeros_i8`
   --> $DIR/leading_zeros.rs:44:12
    |
44  |     *out = i8::leading_zeros(*buffer);
    |            ^^^^^^^^^^^^^^^^^^^^^^^^^^
note: called by `leading_zeros_i8`
   --> $DIR/leading_zeros.rs:40:8
    |
40  | pub fn leading_zeros_i8(
    |        ^^^^^^^^^^^^^^^^

error: cannot cast between pointer types
       from `*u32`
         to `*u16`
    --> /home/firestar99/.rustup/toolchains/nightly-2024-11-22-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/num/mod.rs:1151:5
     |
1151 | /     uint_impl! {
1152 | |         Self = u16,
1153 | |         ActualT = u16,
1154 | |         SignedT = i16,
...    |
1168 | |         bound_condition = "",
1169 | |     }
     | |_____^
     |
note: used from within `<u16>::leading_zeros`
    --> /home/firestar99/.rustup/toolchains/nightly-2024-11-22-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/num/mod.rs:1151:5
     |
1151 | /     uint_impl! {
1152 | |         Self = u16,
1153 | |         ActualT = u16,
1154 | |         SignedT = i16,
...    |
1168 | |         bound_condition = "",
1169 | |     }
     | |_____^
note: called by `<i16>::leading_zeros`
    --> /home/firestar99/.rustup/toolchains/nightly-2024-11-22-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/num/mod.rs:359:5
     |
359  | /     int_impl! {
360  | |         Self = i16,
361  | |         ActualT = i16,
362  | |         UnsignedT = u16,
...    |
377  | |         bound_condition = "",
378  | |     }
     | |_____^
note: called by `leading_zeros::leading_zeros_i16`
    --> $DIR/leading_zeros.rs:52:12
     |
52   |     *out = i16::leading_zeros(*buffer);
     |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^
note: called by `leading_zeros_i16`
    --> $DIR/leading_zeros.rs:48:8
     |
48   | pub fn leading_zeros_i16(
     |        ^^^^^^^^^^^^^^^^^

error: cannot cast between pointer types
       from `*u32`
         to `*u64`
    --> /home/firestar99/.rustup/toolchains/nightly-2024-11-22-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/num/mod.rs:1223:5
     |
1223 | /     uint_impl! {
1224 | |         Self = u64,
1225 | |         ActualT = u64,
1226 | |         SignedT = i64,
...    |
1240 | |         bound_condition = "",
1241 | |     }
     | |_____^
     |
note: used from within `<u64>::leading_zeros`
    --> /home/firestar99/.rustup/toolchains/nightly-2024-11-22-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/num/mod.rs:1223:5
     |
1223 | /     uint_impl! {
1224 | |         Self = u64,
1225 | |         ActualT = u64,
1226 | |         SignedT = i64,
...    |
1240 | |         bound_condition = "",
1241 | |     }
     | |_____^
note: called by `<i64>::leading_zeros`
    --> /home/firestar99/.rustup/toolchains/nightly-2024-11-22-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/num/mod.rs:407:5
     |
407  | /     int_impl! {
408  | |         Self = i64,
409  | |         ActualT = i64,
410  | |         UnsignedT = u64,
...    |
425  | |         bound_condition = "",
426  | |     }
     | |_____^
note: called by `leading_zeros::leading_zeros_i64`
    --> $DIR/leading_zeros.rs:68:12
     |
68   |     *out = i64::leading_zeros(*buffer);
     |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^
note: called by `leading_zeros_i64`
    --> $DIR/leading_zeros.rs:64:8
     |
64   | pub fn leading_zeros_i64(
     |        ^^^^^^^^^^^^^^^^^

error: aborting due to 6 previous errors

